### PR TITLE
[Feature] Add TensorClassModuleBase for type-safe TensorClass modules

### PR DESCRIPTION
**Pull request by az0uz**
## Description

This PR introduces `TensorClassModuleBase`, a new base class that provides a type-safe way to define PyTorch modules that operate on `TensorClass` instances. This implementation offers compile-time type checking and seamless integration with TensorDict-based workflows.

### Key Features:

- **Type-safe module definition**: Generic base class `TensorClassModuleBase[InputClass, OutputClass]` with automatic type extraction
- **TensorDict integration**: `TensorClassModuleWrapper` enables conversion to `TensorDictModule` via `as_td_module()` method
- **Nested TensorClass support**: Automatically handles nested TensorClass structures and extracts all keys
- **ONNX export compatibility**: Full support for ONNX export with proper input/output handling

### Implementation Details:

- New module: `tensordict/nn/tensorclass_module.py` with `TensorClassModuleBase` and `TensorClassModuleWrapper`
- Automatic key extraction from TensorClass type annotations using `_tensor_class_keys()`
- Generic type parameter extraction via `__init_subclass__` for clean API
- Comprehensive test suite in `test/test_tensorclass_module.py` covering forward pass, TensorDict conversion, ONNX export, and edge cases

### Additional Changes:
- Exported new classes in `tensordict/__init__.py` and `tensordict/nn/__init__.py`

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/pytorch/tensordict/issues) for new features and bug fixes)

implements https://github.com/pytorch/tensordict/issues/1355

This change addresses the need for type-safe module definitions when working with TensorClass instances. Previously, developers had to work directly with TensorDict or manually handle type conversions, which was error-prone and lacked compile-time type checking.

With `TensorClassModuleBase`, developers can:
1. Define modules with explicit input/output TensorClass types
2. Get IDE autocomplete and type checking for TensorClass fields
3. Seamlessly integrate with existing TensorDict-based workflows
4. Export to ONNX while maintaining type safety

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/pytorch/tensordict/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have updated the documentation accordingly (via comprehensive docstrings and examples).

### CODE REVIEW

File: tensordict/nn/tensorclass_module.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+        Returns:
+            TensorDict: Output tensordict.
+
+        """
+        return self.tc_module(
+            self.tc_module.input_type.from_dict(
+                input_dict=tensordict.to_dict(),
+                batch_size=tensordict.batch_size,
+                device=tensordict.device,
+            )
```
Comment (vmoens) [inline tensordict/nn/tensorclass_module.py]: can't we use `TensorClass.from_tensordict` directly?
The issue I see here is for example that a tensorclass can contain a LazyStackedTensorDict with exclusive keys (keys present in some but not all elements of the stack). `to_dict` will fail for these. It's also slow.

File: test/test_tensorclass_module.py
Comment (vmoens) [inline test/test_tensorclass_module.py]: I'm not against a new file but in general we've been putting these in `test_nn.py`

**Review by vmoens — COMMENTED** [review]
Thanks for this!
Overall great feature!
I left some comments here and there but they're pretty minor.
Can you sign the CLA? It's required to run the tests and merge

**Comment by az0uz** [comment]
The person responsible for the existing cla in my company has asked cla@ to add me, but haven't got any answer yet, I'll ping them again. Do I need to also sign it personally?

**Comment by vmoens** [comment]
> The person responsible for the existing cla in my company has asked cla@ to add me, but haven't got any answer yet, I'll ping them again. Do I need to also sign it personally?

No the company one should be fine. If we need to ping anyone on the Meta side LMK!

**Comment by az0uz** [comment]
> If we need to ping anyone on the Meta side LMK!

Yes that would be great, no answer for a week, we've tried to ping cla@meta.com again yesterday but still no news.

File: test/test_compile.py [hunk 2]
```diff
@@ -950,9 +950,7 @@ def test_onnx_export_module(self, tmpdir):
         x = torch.randn(3)
         y = torch.randn(3)
         torch_input = {"x": x, "y": y}
-        onnx_program = torch.onnx.dynamo_export(tdm, **torch_input)
-
-        onnx_input = onnx_program.adapt_torch_inputs_to_onnx(**torch_input)
+        onnx_program = torch.onnx.export(tdm, kwargs=torch_input, dynamo=True)
```
[the remark this answers is not in the document]
Comment (az0uz) [reply test/test_compile.py]: the onnx fix standalone is available here: https://github.com/pytorch/tensordict/pull/1478

File: tensordict/nn/tensorclass_module.py
[the remark this answers is not in the document]
Comment (az0uz) [reply tensordict/nn/tensorclass_module.py]: Do I need to write anything else than adding them to `docs/source/reference/nn.rst` like above?

**Comment by vmoens** [comment]
GHA seems down -- we'll need to wait till tests run for this

File: tensordict/nn/tensorclass_module.py
[the remark this answers is not in the document]
Comment (vmoens) [reply tensordict/nn/tensorclass_module.py]: That's really cool thx
